### PR TITLE
Snowbridge - Fix benchmarks, add fee multipler and move config to Ethereum module

### PR DESCRIPTION
**Pull request by claravanstaden**
The Snowbridge Inbound Queue pallet benchmarks are failing:

```
2024-03-21 13:50:40 Running  benchmark: snowbridge_pallet_ethereum_client.submit_execution_header(0 args) 1/1 1/1    
Running benchmarks for snowbridge_pallet_inbound_queue to ./bridge-hub-kusama-weights/
2024-03-21 13:50:41 Starting benchmark: snowbridge_pallet_inbound_queue::submit    
2024-03-21 13:50:41 💫 Verifying message with block hash 0x3921…3a3f    
2024-03-21 13:50:41 💫 Receipt verification successful for 0x3921…3a3f    
2024-03-21 13:50:41 panicked at /home/bparity/.cargo/registry/src/index.crates.io-6f17d22bba15001f/snowbridge-pallet-inbound-queue-0.1.0/src/benchmarking/mod.rs:43:13:
Expected Ok(_). Got Err(
    Module(
        ModuleError {
            index: 80,
            error: [
                0,
                0,
                0,
                0,
            ],
            message: Some(
                "InvalidGateway",
            ),
        },
    ),
)    
Error: 
   0: Invalid input: Error executing and verifying runtime benchmark: Execution aborted due to trap: wasm trap: wasm `unreachable` instruction executed
      WASM backtrace:
      error while executing at wasm backtrace:
          0: 0x436fa6 - <unknown>!rust_begin_unwind
          1: 0x2969 - <unknown>!core::panicking::panic_fmt::hbb0c5dd9c7ebab99
          2: 0x329f54 - <unknown>!core::ops::function::FnOnce::call_once{{vtable.shim}}::h0374b352d74ea114
          3: 0x300e24 - <unknown>!<bridge_hub_kusama_runtime::Runtime as frame_benchmarking::utils::runtime_decl_for_benchmark::BenchmarkV1<sp_runtime::generic::block::Block<sp_runtime::generic::header::Header<u32,sp_runtime::traits::BlakeTwo256>,sp_runtime::generic::unchecked_extrinsic::UncheckedExtrinsic<sp_runtime::multiaddress::MultiAddress<<<sp_runtime::MultiSignature as sp_runtime::traits::Verify>::Signer as sp_runtime::traits::IdentifyAccount>::AccountId,()>,bridge_hub_kusama_runtime::RuntimeCall,sp_runtime::MultiSignature,(frame_system::extensions::check_non_zero_sender::CheckNonZeroSender<bridge_hub_kusama_runtime::Runtime>,frame_system::extensions::check_spec_version::CheckSpecVersion<bridge_hub_kusama_runtime::Runtime>,frame_system::extensions::check_tx_version::CheckTxVersion<bridge_hub_kusama_runtime::Runtime>,frame_system::extensions::check_genesis::CheckGenesis<bridge_hub_kusama_runtime::Runtime>,frame_system::extensions::check_mortality::CheckMortality<bridge_hub_kusama_runtime::Runtime>,frame_system::extensions::check_nonce::CheckNonce<bridge_hub_kusama_runtime::Runtime>,frame_system::extensions::check_weight::CheckWeight<bridge_hub_kusama_runtime::Runtime>,pallet_transaction_payment::ChargeTransactionPayment<bridge_hub_kusama_runtime::Runtime>,bridge_hub_kusama_runtime::BridgeRejectObsoleteHeadersAndMessages,bridge_runtime_common::refund_relayer_extension::RefundSignedExtensionAdapter<bridge_runtime_common::refund_relayer_extension::RefundBridgedParachainMessages<bridge_hub_kusama_runtime::Runtime,bridge_runtime_common::refund_relayer_extension::RefundableParachain<frame_support::instances::Instance1,bp_bridge_hub_polkadot::BridgeHubPolkadot>,bridge_runtime_common::refund_relayer_extension::RefundableMessagesLane<frame_support::instances::Instance1,bridge_hub_kusama_runtime::bridge_to_polkadot_config::AssetHubKusamaToAssetHubPolkadotMessagesLane>,bridge_runtime_common::refund_relayer_extension::ActualFeeRefund<bridge_hub_kusama_runtime::Runtime>,bridge_hub_kusama_runtime::bridge_to_polkadot_config::PriorityBoostPerMessage,bridge_hub_kusama_runtime::bridge_to_polkadot_config::StrRefundBridgeHubPolkadotMessages>>)>>>>::dispatch_benchmark::h500c8f48422e47bd
          4: 0x3c0b6a - <unknown>!Benchmark_dispatch_benchmark
```

This PR fixes the error (the Ethereum gateway contract address should be set in the benchmarks since it is verified in the inbound queue when a message is submitted) and moves the benchmark config into `bridge_to_ethereum_config.rs`.

- [x] Does not require a CHANGELOG entry

### CODE REVIEW

File: system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_config.rs [hunk 1]
```diff
@@ -25,3 +25,43 @@ pub type SnowbridgeExporter = EthereumBlobExporter<
 	snowbridge_pallet_outbound_queue::Pallet<Runtime>,
 	snowbridge_core::AgentIdOf,
 >;
+
```
Comment (bkontur) [inline system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_config.rs:28]: @claravanstaden can you please move all your pallet inits and `parameter_types! {` to this file from libs.rs?

File: system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs [hunk 2]
```diff
@@ -25,3 +25,42 @@ pub type SnowbridgeExporter = EthereumBlobExporter<
 	snowbridge_pallet_outbound_queue::Pallet<Runtime>,
 	snowbridge_core::AgentIdOf,
 >;
+
+#[cfg(feature = "runtime-benchmarks")]
+pub mod benchmark_helpers {
+	use crate::{EthereumBeaconClient, Runtime, RuntimeOrigin};
```
Comment (bkontur) [inline system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs:31]:
```suggestion
	use crate::{EthereumBeaconClient, EthereumGatewayAddress, Runtime, RuntimeOrigin};
```

File: system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_config.rs [hunk 3]
```diff
@@ -25,3 +25,43 @@ pub type SnowbridgeExporter = EthereumBlobExporter<
...
+#[cfg(feature = "runtime-benchmarks")]
+pub mod benchmark_helpers {
+	use crate::{EthereumBeaconClient, Runtime, RuntimeOrigin};
```
Comment (bkontur) [inline system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_config.rs:31]:
```suggestion
	use crate::{EthereumBeaconClient, EthereumGatewayAddress, Runtime, RuntimeOrigin};
```

File: system-parachains/bridge-hubs/bridge-hub-polkadot/src/lib.rs [hunk 4]
```diff
@@ -500,22 +499,17 @@ impl pallet_utility::Config for Runtime {
...
 }
 
-// Ethereum Bridge
-parameter_types! {
-	// The gateway address is set by governance.
-	pub storage EthereumGatewayAddress: H160 = H160::zero();
-}
-
+// Ethereum Bridge pallets
 impl snowbridge_pallet_inbound_queue::Config for Runtime {
```
Comment (bkontur) [inline system-parachains/bridge-hubs/bridge-hub-polkadot/src/lib.rs:503]: @claravanstaden I meant everything and pallets setup also :), everything related to the Eth, move to the `bridge_to_ethereum_config.rs` please :)

**Comment by acatangiu** [comment]
https://github.com/Snowfork/runtimes/pull/5 for bringing in latest fixes

File: system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs (hunk 2 above)
Comment (claravanstaden) [reply system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs]: I moved things around, please let me know if it looks OK.

File: system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_config.rs (hunk 3 above)
Comment (claravanstaden) [reply system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_config.rs]: I moved things around, please let me know if it looks OK.

Comment (bkontur) [reply system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_config.rs]: @claravanstaden nice, thank you, just please update description and mark ready for review

**Review by bkontur — APPROVED** [review]

**Review by acatangiu — APPROVED** [review]

**Comment by claravanstaden** [comment]
@bkontur @acatangiu I think need to fix the failing bridge-hub-kusama integration test quick before we merge.

File: system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs [hunk 5]
```diff
@@ -25,3 +38,183 @@ pub type SnowbridgeExporter = EthereumBlobExporter<
...
+	type MessageConverter = MessageToXcm<
+		CreateAssetCall,
+		bp_asset_hub_kusama::CreateForeignAssetDeposit,
+		InboundQueuePalletInstance,
+		AccountId,
+		Balance,
+	>;
+	type WeightToFee = WeightToFee;
+	type LengthToFee = ConstantMultiplier<Balance, TransactionByteFee>;
+	type MaxMessageSize = ConstU32<2048>;
```
Comment (eskimor) [inline system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs:68]: Magic number? At least some docs/comment about the reasoning would be good.

File: system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs [hunk 6]
```diff
@@ -25,3 +38,183 @@ pub type SnowbridgeExporter = EthereumBlobExporter<
...
+	type PricingParameters = EthereumSystem;
+	type AssetTransactor = <xcm_config::XcmConfig as xcm_executor::Config>::AssetTransactor;
+}
+
+impl snowbridge_pallet_outbound_queue::Config for Runtime {
+	type RuntimeEvent = RuntimeEvent;
+	type Hashing = Keccak256;
+	type MessageQueue = MessageQueue;
+	type Decimals = ConstU8<12>;
+	type MaxMessagePayloadSize = ConstU32<2048>;
```
Comment (eskimor) [inline system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs:79]: If this has to be the same as the `MaxMessageSize`, maybe a constant would be good?

File: system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs [hunk 7]
```diff
@@ -25,3 +38,183 @@ pub type SnowbridgeExporter = EthereumBlobExporter<
...
+	type MaxMessagesPerBlock = ConstU32<32>;
```
Comment (eskimor) [inline system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs:80]: Another magic number.

File: system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs [hunk 8]
```diff
@@ -25,3 +38,183 @@ pub type SnowbridgeExporter = EthereumBlobExporter<
...
+		},
+		deneb: Fork {
+			version: [4, 0, 0, 0], // 0x04000000
+			epoch: 0,
+		}
+	};
+}
+
+parameter_types! {
+	pub const MaxExecutionHeadersToKeep: u32 = 8192 * 20;
```
Comment (eskimor) [inline system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs:142]: Why that number?

File: system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_config.rs [hunk 9]
```diff
@@ -25,3 +38,188 @@ pub type SnowbridgeExporter = EthereumBlobExporter<
...
+impl snowbridge_pallet_system::Config for Runtime {
+	type RuntimeEvent = RuntimeEvent;
+	type OutboundQueue = EthereumOutboundQueue;
+	type SiblingOrigin = EnsureXcm<AllowSiblingsOnly>;
+	type AgentIdOf = snowbridge_core::AgentIdOf;
+	type TreasuryAccount = TreasuryAccount;
+	type Token = Balances;
+	type WeightInfo = crate::weights::snowbridge_pallet_system::WeightInfo<Runtime>;
+	#[cfg(feature = "runtime-benchmarks")]
+	type Helper = ();
```
Comment (bkchr) [inline system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_config.rs:161]: Why is this not set?

File: system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs [hunk 10]
```diff
@@ -25,3 +38,188 @@ pub type SnowbridgeExporter = EthereumBlobExporter<
...
+impl snowbridge_pallet_system::Config for Runtime {
+	type RuntimeEvent = RuntimeEvent;
+	type OutboundQueue = EthereumOutboundQueue;
+	type SiblingOrigin = EnsureXcm<AllowSiblingsOnly>;
+	type AgentIdOf = snowbridge_core::AgentIdOf;
+	type TreasuryAccount = TreasuryAccount;
+	type Token = Balances;
+	type WeightInfo = crate::weights::snowbridge_pallet_system::WeightInfo<Runtime>;
+	#[cfg(feature = "runtime-benchmarks")]
+	type Helper = ();
```
Comment (bkchr) [inline system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs:161]: Same

**Review by bkchr — APPROVED** [review]

**Review by eskimor — COMMENTED** [review]
Looks good, left a few nits/remarks.

**Review by eskimor — APPROVED** [review]

File: system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_config.rs (hunk 9 above)
Comment (claravanstaden) [reply system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_config.rs]: Fixed here: https://github.com/polkadot-fellows/runtimes/pull/252/commits/73ba2ba4b28fba0934ea3faa3ee1b19fb5356893

File: system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs (hunk 10 above)
Comment (claravanstaden) [reply system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs]: Fixed here: https://github.com/polkadot-fellows/runtimes/pull/252/commits/73ba2ba4b28fba0934ea3faa3ee1b19fb5356893

File: system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs (hunk 8 above)
Comment (claravanstaden) [reply system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs:145]: Added a comment here: https://github.com/polkadot-fellows/runtimes/pull/252/commits/410f9c31b259f71540b49a92d06726f5c3c012d8

File: system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_config.rs (hunk 9 above)
Comment (bkchr) [reply system-parachains/bridge-hubs/bridge-hub-polkadot/src/bridge_to_ethereum_config.rs]: Ty!

**Review by vgeddes — APPROVED** [review]

File: system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs (hunk 7 above)
Comment (vgeddes) [reply system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs:80]: Agreed, we should make this a documented comment. `MaxMessagesPerBlock` refers to the maximum number of outbound messages to Ethereum that can be committed per block.

File: system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs (hunk 6 above)
Comment (vgeddes) [reply system-parachains/bridge-hubs/bridge-hub-kusama/src/bridge_to_ethereum_config.rs:79]: Definitely agree that the magic numbers needs to be hoisted into documented constants. Though `MaxMessagePayloadSize` does have a different purpose to `snowbridge_pallet_inbound_queue::MaxMessageSize`.

* inbound_queue::MaxMessageSize: Worst case message size (including proofs), used to calculate refunds for relayers.
* outbound_queue::MaxMessagePayloadSize: Maximum allowed size of encoded message payload that is committed for delivery to Ethereum

**Comment by bkchr** [comment]
@claravanstaden tests are now failing. Are you working on fixing this?

**Comment by bkontur** [comment]
@claravanstaden @vgeddes my guess is that, these fee amounts [here](https://github.com/polkadot-fellows/runtimes/pull/252/files#diff-bd26e40b81351dde58a2a2ef23c118cf04731f18bcda5ce2961b8df25026a5f3R575) and [here](https://github.com/polkadot-fellows/runtimes/pull/252/files#diff-46a0e8ad1ccebb1fd2c05ed3202f1d08efec5f64c55fafc3504ca8ec462592f0R579) are weight-dependent, I would maybe remove those  amount checks e.g. `&& *amount == 50710000`, because we will have to adjust them on every runtime upgrade or weights change.

(But checking locally, if it is about that)

**Comment by vgeddes** [comment]
> @claravanstaden tests are now failing. Are you working on fixing this?

About to push a fix, just waiting on tests to run..

**Comment by bkontur** [comment]
> @claravanstaden @vgeddes my guess is that, these fee amounts [here](https://github.com/polkadot-fellows/runtimes/pull/252/files#diff-bd26e40b81351dde58a2a2ef23c118cf04731f18bcda5ce2961b8df25026a5f3R575) and [here](https://github.com/polkadot-fellows/runtimes/pull/252/files#diff-46a0e8ad1ccebb1fd2c05ed3202f1d08efec5f64c55fafc3504ca8ec462592f0R579) are weight-dependent, I would maybe remove those amount checks e.g. `&& *amount == 50710000`, because we will have to adjust them on every runtime upgrade or weights change.
> 
> (But checking locally, if it is about that)

here is my suggestion with working tests: https://github.com/Snowfork/runtimes/pull/8

**Comment by ggwpez** [comment]
We dont have a CI check for benchmarks? We have some on the SDK, should probably have them here as well.

**Comment by bkchr** [comment]
@vgeddes please run `cargo fmt --all`

**Comment by acatangiu** [comment]
migration CI jobs were automatically cancelled for no apparent reason 🤷 let's see if it goes ahead with the merge or needs manual intervention from some gh polkadot-fellows org member

**Pull request merged**